networkx/algorithms/community/modularity_max.py

```diff
@@ -234,7 +215,98 @@ def greedy_modularity_communities(G, weight=None, resolution=1, n_communities=1)
+
+
+def greedy_modularity_communities(
+    G, weight=None, resolution=1, cutoff=1, best_n=None, **aliases
+):
+    """
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    weight : string or None, optional (default=None)
+        The name of an edge attribute that holds the numerical value used
+        as a weight.  If None, then each edge has weight 1.
+        The degree is the sum of the edge weights adjacent to the node.
+
+    resolution : float, optional (default=1)
+        If resolution is less than 1, modularity favors larger communities.
+        Greater than 1 favors smaller communities.
+
+    cutoff : int or None, optional (default=1)
+        The number of communities in the partition at which the merging process
+        will stop even if a maximum is not yet reached.
+
+    best_n : int or None, optional (default=None)
+        Force community merging to continue until `best_n` communities remain
+        even if the modularity decreases.
+        If ``None``, don't force it to continue beyond a maximum.
+
+    Returns
+    -------
+    communities: list
+        A list of frozensets of nodes, one for each community.
+        Sorted by length with largest communities first.
+
+    Examples
+    --------
+    >>> from networkx.algorithms.community import greedy_modularity_communities
+    >>> G = nx.karate_club_graph()
+    >>> c = greedy_modularity_communities(G)
+    >>> sorted(c[0])
+    [8, 14, 15, 18, 20, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]
+
+    See Also
+    --------
+    modularity
+    """
+    if (cutoff < 1) or (cutoff > G.number_of_nodes()):
+        raise ValueError(
+            f"``cutoff`` must be between 1 and {G.number_of_nodes()}. Got {cutoff}."
+        )
+    if best_n is not None:
+        if (best_n < 1) or (best_n > G.number_of_nodes()):
+            raise ValueError(
+                f"``best_n`` must be between 1 and {G.number_of_nodes()}. Got {best_n}."
+            )
+        if best_n < cutoff:
+            raise ValueError(
+                f"``best_n`` (got {best_n}) must be greater or equal than ``cutoff`` "
+                f"(got {cutoff})."
+            )
+    else:
+        best_n = G.number_of_nodes()
+    if "n_communities" in aliases:
+        import warnings
+
+        warnings.warn(
+            "kwarg ``n_communities`` in greedy_modularity_communities is deprecated"
+            "and will be removed in version 3.0.   Use ``cutoff`` instead.",
+            DeprecationWarning,
+        )
+        cutoff = aliases.pop("n_communities")
+
+    # retrieve generator object to construct output
+    community_gen = _greedy_modularity_communities_generator(
+        G, weight=weight, resolution=resolution
+    )
+
+    # construct the first best community
+    communities = next(community_gen)
+    if len(communities) == 1:
+        return communities
+    dq = next(community_gen)
+
+    while len(communities) > cutoff and not (dq < 0 and len(communities) <= best_n):
```

Comment:
Why should the condition involving `best_n` involve `dq` at all.  The doc_string says `best_n` makes the algorithm stop when you have n communities "even if modularity decreases". But the code implements "only if modularity has started to decrease.       

How does the user make it stop at exact n communities?

It seems like we've got too many different possibilities here. You can stop:
1) when modularity is at a max, 
2) when n == cutoff and no max found
3) when n == best_n and max found
[other possibilities]
4) when n == best_n independent of whether max has been found.

Is there a simple way to do this?

Reply:
getting confused now, too... if i understand your question correctly: if you want to stop at an exact `n` communities, you can set `cutoff` equal to `best_n` equal to `n`, ie `greedy_modularity_communities(G, cutoff=n, best_n=n)`.